File: src/org/andstatus/app/context/MyPreferences.java
/**
 * Copyright (C) 2013 yvolk (Yuri Volkov), http://yurivolkov.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.andstatus.app.context;

import org.andstatus.app.TimelineActivity;
import org.andstatus.app.util.MyLog;

import android.app.backup.BackupManager;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.PackageManager.NameNotFoundException;
import android.os.Environment;
import android.preference.PreferenceManager;
import android.text.TextUtils;

import java.io.File;

/**
 * This is a central point of accessing SharedPreferences
 * Noninstantiable class 
 * @author yvolk@yurivolkov.com
 */
public class MyPreferences {
    private static final String TAG = MyPreferences.class.getSimpleName();
    
    /**
     * This is sort of button to start verification of credentials
     */
    public static final String KEY_VERIFY_CREDENTIALS = "verify_credentials";

    public static final String KEY_HISTORY_SIZE = "history_size";
    public static final String KEY_HISTORY_TIME = "history_time";
    /**
     * Period of automatic updates in seconds
     */
    public static final String KEY_SYNC_FREQUENCY_SECONDS = "fetch_frequency";
    public static final String KEY_SYNC_INDICATOR_ON_TIMELINE = "sync_indicator_on_timeline";
    public static final String KEY_SYNC_WHILE_USING_APPLICATION = "sync_while_using_application";
    public static final String KEY_DOWNLOAD_ATTACHMENTS_OVER_WIFI_ONLY = "download_attachments_over_wifi_only";
    public static final String KEY_CONNNECTION_TIMEOUT_SECONDS = "connection_timeout";
    public static final String KEY_RINGTONE_PREFERENCE = "notification_ringtone";
    public static final String KEY_CONTACT_DEVELOPER = "contact_developer";
    public static final String KEY_REPORT_BUG = "report_bug";
    public static final String KEY_CHANGE_LOG = "change_log";
    public static final String KEY_ABOUT_APPLICATION = "about_application";

    /**
     * System time when shared preferences were changed
     */
    public static final String KEY_PREFERENCES_CHANGE_TIME = "preferences_change_time";
    public static final String KEY_DATA_PRUNED_DATE = "data_pruned_date";
    /**
     * Minimum logging level for the whole application (i.e. for any tag)
     */
    public static final String KEY_MIN_LOG_LEVEL = "min_log_level";
    public static final String KEY_SENDING_MESSAGES_LOG_ENABLED = "sending_messages_log_enabled";
    
    public static final String KEY_THEME_SIZE = "theme_size";
    public static final String KEY_THEME_COLOR = "theme_color";
    public static final String KEY_SHOW_AVATARS = "show_avatars";
    public static final String KEY_SHOW_ATTACHED_IMAGES = "show_attached_images";

    /**
     * Use this dir: http://developer.android.com/reference/android/content/Context.html#getExternalFilesDir(java.lang.String)
     * (for API 8)
     */
    public static final String KEY_USE_EXTERNAL_STORAGE = "use_external_storage";
    /**
     * New value for #KEY_USE_EXTERNAL_STORAGE to e confirmed/processed
     */
    public static final String KEY_USE_EXTERNAL_STORAGE_NEW = "use_external_storage_new";
    public static final String KEY_ENABLE_ANDROID_BACKUP = "enable_android_backup";
    /**
     * Is the timeline combined in {@link TimelineActivity} 
     */
    public static final String KEY_TIMELINE_IS_COMBINED = "timeline_is_combined";
    /**
     * Version code of last opened application (int) 
     */
    public static final String KEY_VERSION_CODE_LAST = "version_code_last";

    public static final String KEY_NOTIFY_OF_COMMANDS_IN_THE_QUEUE = "notifications_queue";
	public static final String KEY_ENTER_SENDS_MESSAGE = "enter_sends_message";
    public static final String KEY_OLD_MESSAGES_FIRST_IN_CONVERSATION = "old_messages_first_in_conversation";
    public static final String KEY_SYNC_AFTER_MESSAGE_WAS_SENT = "sync_after_message_was_sent";
    public static final String KEY_MARK_REPLIES_IN_TIMELINE = "mark_replies_in_timeline";
	
    private MyPreferences(){
        throw new AssertionError();
    }
    
    private static volatile boolean defaultSharedPreferencesLocked = false;
    public static SharedPreferences getDefaultSharedPreferences() {
        if (defaultSharedPreferencesLocked) {
            return null;
        }
        Context context = MyContextHolder.get().context();
        if (context == null) {
            MyLog.e(TAG, "getDefaultSharedPreferences - Was not initialized yet");
            for(StackTraceElement element : Thread.currentThread().getStackTrace()) { 
                MyLog.v(TAG, element.toString()); 
            }
            return null;
        } else {
            return PreferenceManager.getDefaultSharedPreferences(context);
        }
    }

    public static void lockDefaultSharedPreferences() {
        defaultSharedPreferencesLocked = true;
    }

    public static void unlockDefaultSharedPreferences() {
        defaultSharedPreferencesLocked = false;
    }
    
    public static boolean shouldSetDefaultValues() {
        SharedPreferences sp = getSharedPreferences(PreferenceManager.KEY_HAS_SET_DEFAULT_VALUES);
        if (sp == null) {
            return false;
        } else {
            boolean areSetAlready = sp
                    .getBoolean(PreferenceManager.KEY_HAS_SET_DEFAULT_VALUES, false);
            return !areSetAlready;
        }
    }
    
    public static void setDefaultValues(int resId, boolean readAgain) {
        Context context = MyContextHolder.get().context();
        if (context == null) {
            MyLog.e(TAG, "setDefaultValues - Was not initialized yet");
        } else {
            PreferenceManager.setDefaultValues(context, resId, readAgain);
        }
    }
    
    public static SharedPreferences getSharedPreferences(String name) {
        Context context = MyContextHolder.get().context();
        if (context == null) {
            MyLog.e(TAG, "getSharedPreferences for " + name + " - were not initialized yet");
            for(StackTraceElement element : Thread.currentThread().getStackTrace()) { 
                MyLog.v(TAG, element.toString()); 
            }
            return null;
        } else {
            return context.getSharedPreferences(name, android.content.Context.MODE_PRIVATE);
        }
    }

    private static final long SYNC_FREQUENCY_DEFAULT_SECONDS = 180;
    /**
     * @return the number of seconds between two sync ("fetch"...) actions.
     */
    public static long getSyncFrequencySeconds() {
        return getLongStoredAsString(KEY_SYNC_FREQUENCY_SECONDS, SYNC_FREQUENCY_DEFAULT_SECONDS);
    }

    private static long getLongStoredAsString(String key, long defaultValue) {
        long longValue = defaultValue;
        SharedPreferences sp = getDefaultSharedPreferences();
        if (sp != null) {
            try {
                long longValueStored = Long.parseLong(sp.getString(key, "0"));
                if (longValueStored > 0) { 
                    longValue = longValueStored;
                }
            } catch (NumberFormatException e) {
                MyLog.v(TAG, e);
            }
        }
        return longValue;
    }

    private static final long CONNNECTION_TIMEOUT_DEFAULT_SECONDS = 30;
    public static int getConnectionTimeoutMs() {
        return (int) java.util.concurrent.TimeUnit.SECONDS.toMillis(getLongStoredAsString(
                KEY_CONNNECTION_TIMEOUT_SECONDS, CONNNECTION_TIMEOUT_DEFAULT_SECONDS));
    }
    
    /**
     * @return the number of milliseconds between two sync ("fetch"...) actions.
     */
    public static long getSyncFrequencyMs() {
        return java.util.concurrent.TimeUnit.SECONDS.toMillis(getSyncFrequencySeconds());
    }
    
    /**
     *  Event: Preferences have changed right now
     *  Remember when last changes to the preferences were made
     */
    public static void onPreferencesChanged() {
        putLong(KEY_PREFERENCES_CHANGE_TIME, System.currentTimeMillis());
        Context context = MyContextHolder.get().context();
        if (context != null && getBoolean(KEY_ENABLE_ANDROID_BACKUP, false)) {
            new BackupManager(context).dataChanged();
        }
    }

	public static void putLong(String key, long value) {
        SharedPreferences sp = getDefaultSharedPreferences();
        if (sp != null) {
            sp.edit().putLong(key, value).commit();
        }
    }
	
    /**
     * @return System time when AndStatus preferences were last time changed. 
     * We take into account here time when accounts were added/removed...
     */
    public static long getPreferencesChangeTime() {
		return getLong(KEY_PREFERENCES_CHANGE_TIME);
    }
    
    public static long getLong(String key) {
        long value = 0;
        SharedPreferences sp = getDefaultSharedPreferences();
        if (sp != null) {
            value = getDefaultSharedPreferences().getLong(key, 0);            
        }
        return value;
    }
	
	/**
     * Standard directory in which to place databases
     */
    public static final String DIRECTORY_DATABASES = "databases";
    public static final String DIRECTORY_DOWNLOADS = "downloads";
    
    /**
     * This function works just like {@link android.content.Context#getExternalFilesDir
     * Context.getExternalFilesDir}, but it takes {@link #KEY_USE_EXTERNAL_STORAGE} into account,
     * so it returns directory either on internal or external storage.
     * 
     * @param type The type of files directory to return.  May be null for
     * the root of the files directory or one of
     * the following Environment constants for a subdirectory:
     * {@link android.os.Environment#DIRECTORY_PICTURES Environment.DIRECTORY_...} (since API 8),
     * {@link #DIRECTORY_DATABASES}
     * @param forcedUseExternalStorage if not null, use this value instead of stored in preferences as {@link #KEY_USE_EXTERNAL_STORAGE}
     * 
     * @return directory, already created for you OR null in case of error
     * @see <a href="http://developer.android.com/guide/topics/data/data-storage.html#filesExternal">filesExternal</a>
     */
    public static File getDataFilesDir(String type, Boolean forcedUseExternalStorage) {
        return getDataFilesDir(type, forcedUseExternalStorage, true);
    }

    public static File getDataFilesDir(String type, Boolean forcedUseExternalStorage, boolean logged) {
        final String method = "getDataFilesDir";
        boolean logEnabled = false;
        File dir = null;
        StringBuilder textToLog = new StringBuilder();
        MyContext myContext = MyContextHolder.get();
        if (myContext.context() == null) {
            textToLog.append("No android.content.Context yet");
        } else {
            if (isStorageExternal(forcedUseExternalStorage)) {
                if (isWritableExternalStorageAvailable(textToLog)) {
                    dir = myContext.context().getExternalFilesDir(type);
                }
            } else {
                dir = myContext.context().getFilesDir();
                if (!TextUtils.isEmpty(type)) {
                    dir = new File(dir, type);
                }
            }
            if (dir != null && !dir.exists()) {
                try {
                    dir.mkdirs();
                } catch (Exception e) {
                    if (logged) {
                        MyLog.e(TAG, method + "; Error creating directory", e);
                    }
                } finally {
                    if (!dir.exists()) {
                        textToLog.append("Could not create '" + dir.getPath() + "'");
                        dir = null;
                    }
                }
            }
            if (logged && logEnabled && MyLog.isLoggable(TAG, MyLog.VERBOSE)) {
                MyLog.v(TAG, method + "; " + (isStorageExternal(forcedUseExternalStorage) ? "External" : "Internal") 
                        + " path: '" + ( (dir == null) ? "(null)" : dir ) + "'");
            }
        }
        if (logged && textToLog.length() > 0) {
            MyLog.i(TAG, method + "; " + textToLog);
        }
        return dir;
    }

    public static boolean isWritableExternalStorageAvailable(StringBuilder textToLog) {
        String state = Environment.getExternalStorageState();
        boolean available = false;
        if (Environment.MEDIA_MOUNTED.equals(state)) {
            available = true;
        } else if (Environment.MEDIA_MOUNTED_READ_ONLY.equals(state)) {
            if (textToLog != null) {
                textToLog.append("We can only read External storage");
            }
        } else {    
            if (textToLog != null) {
                textToLog.append("Error accessing External storage, state='" + state + "'");
            }
        }
        return available;
    }
    
    public static boolean isStorageExternal(Boolean forcedUseExternalStorage) {
        boolean useExternalStorage = false;
        if (forcedUseExternalStorage == null) {
            useExternalStorage = getBoolean(KEY_USE_EXTERNAL_STORAGE, false); 
        } else {
            useExternalStorage  = forcedUseExternalStorage;
        }
        return useExternalStorage;
    }

    public static boolean isSyncWhileUsingApplicationEnabled() {
        return getBoolean(KEY_SYNC_WHILE_USING_APPLICATION, true);
    }
    
    public static boolean getBoolean(String key, boolean defaultValue) {
        boolean value = defaultValue;
        SharedPreferences sp = getDefaultSharedPreferences();
        if (sp != null) {
            value = getDefaultSharedPreferences().getBoolean(key, defaultValue); 
        }
        return value;
    }
    
    /**
     * Extends {@link android.content.ContextWrapper#getDatabasePath(java.lang.String)}
     * @param name The name of the database for which you would like to get its path.
     * @param forcedUseExternalStorage if not null, use this value instead of stored in preferences as {@link #KEY_USE_EXTERNAL_STORAGE}
     * @return
     */
    public static File getDatabasePath(String name, Boolean forcedUseExternalStorage) {
        File dbDir = getDataFilesDir(MyPreferences.DIRECTORY_DATABASES, forcedUseExternalStorage);
        File dbAbsolutePath = null;
        if (dbDir != null) {
            dbAbsolutePath = new File(dbDir.getPath() + "/" + name);
        }
        return dbAbsolutePath;
    }
    
    /**
     * Simple check that allows to prevent data access errors
     */
    public static boolean isDataAvailable() {
        return getDataFilesDir(null, null) != null;
    }

    /**
     * Load the theme according to the preferences.
     */
    public static void loadTheme(Object tag, Context context) {
        String themeColor = getDefaultSharedPreferences().getString(KEY_THEME_COLOR, "DeviceDefault");
        StringBuilder themeName = new StringBuilder("Theme.");
        themeName.append(themeColor);
        themeName.append(".AndStatus.");
        String themeSize = getDefaultSharedPreferences().getString(KEY_THEME_SIZE, "StandardSize");
        themeName.append(themeSize);
        int themeId = context.getResources().getIdentifier(themeName.toString(), "style",
                "org.andstatus.app");
        if (themeId == 0 || MyLog.isLoggable(tag, MyLog.VERBOSE)) {
            String text = "loadTheme; theme=\"" + themeName.toString() + "\"; id=" + Integer.toHexString(themeId);
            if (themeId == 0) {
                MyLog.e(tag, text);
            } else {
                MyLog.v(tag, text);
            }
        }
        context.setTheme(themeId);
    }
    
    public static boolean isThemeLight() {
       return getDefaultSharedPreferences().getString(KEY_THEME_COLOR, "DeviceDefault").contains("Light");
    }

    /**
     * @param context
     * @param update
     * @return true if we opened previous version
     * @throws NameNotFoundException
     */
    public static boolean checkAndUpdateLastOpenedAppVersion(Context context, boolean update) {
        boolean changed = false;
        int versionCodeLast =  getDefaultSharedPreferences().getInt(KEY_VERSION_CODE_LAST, 0);
        PackageManager pm = context.getPackageManager();
        PackageInfo pi;
        try {
            pi = pm.getPackageInfo(context.getPackageName(), 0);
            int versionCode =  pi.versionCode;
            if (versionCodeLast < versionCode) {
                // Even if the User will see only the first page of the Help activity,
                // count this as showing the Change Log
                MyLog.v(TAG, "Last opened version=" + versionCodeLast + ", current is " + versionCode
                        + (update ? ", updating" : "")
                        );
                changed = true;
                if ( update && MyContextHolder.get().isReady()) {
                    getDefaultSharedPreferences().edit().putInt(KEY_VERSION_CODE_LAST, versionCode).commit();
                }
            }
        } catch (NameNotFoundException e) {
            MyLog.e(TAG, "Unable to obtain package information", e);
        }
        return changed;
    }

    public static boolean showAvatars() {
        return getDefaultSharedPreferences().getBoolean(MyPreferences.KEY_SHOW_AVATARS, true);
    }

    public static boolean showAttachedImages() {
        return getDefaultSharedPreferences().getBoolean(MyPreferences.KEY_SHOW_ATTACHED_IMAGES, true);
    }
}

File: app/src/main/kotlin/org/andstatus/app/context/MyPreferences.kt
/*
 * Copyright (C) 2016 yvolk (Yuri Volkov), http://yurivolkov.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.andstatus.app.context

import android.app.backup.BackupManager
import android.content.Context
import android.net.Uri
import android.os.Build
import org.andstatus.app.R
import org.andstatus.app.timeline.TapOnATimelineTitleBehaviour
import org.andstatus.app.util.SharedPreferencesUtil
import org.andstatus.app.util.UriUtils
import java.util.concurrent.TimeUnit

/**
 * This is a central point of accessing SharedPreferences
 * @author yvolk@yurivolkov.com
 */
object MyPreferences {
    // ----------------------------------------------------------
    // Accounts
    // (no keys/settings here yet)
    // ----------------------------------------------------------
    // Appearance
    val KEY_CUSTOM_LOCALE: String = "custom_locale"
    val KEY_THEME_COLOR: String = "theme_color"
    val KEY_ACTION_BAR_BACKGROUND_COLOR: String = "action_bar_background_color"
    val KEY_ACTION_BAR_TEXT_COLOR: String = "action_bar_text_color"
    val KEY_BACKGROUND_COLOR: String = "background_color"
    val KEY_THEME_SIZE: String = "theme_size"
    val KEY_ROUNDED_AVATARS: String = "rounded_avatars"
    val KEY_SHOW_IMAGE_ANIMATIONS: String = "show_image_animations"

    // ----------------------------------------------------------
    // Timeline
    val KEY_DEFAULT_TIMELINE: String = "default_timeline"
    val KEY_TAP_ON_A_TIMELINE_TITLE_BEHAVIOUR: String = "tap_on_a_timeline_title"
    val KEY_ACTOR_IN_TIMELINE: String = "user_in_timeline"
    val KEY_SHOW_AVATARS: String = "show_avatars"
    val KEY_SHOW_ORIGIN: String = "show_origin"
    val KEY_SHOW_SENSITIVE_CONTENT: String = "show_sensitive"
    val KEY_MARK_REPLIES_TO_ME_IN_TIMELINE: String = "mark_replies_in_timeline"
    val KEY_SHOW_BUTTONS_BELOW_NOTE: String = "show_buttons_below_message"
    val KEY_OLD_NOTES_FIRST_IN_CONVERSATION: String = "old_messages_first_in_conversation"
    val KEY_REFRESH_TIMELINE_AUTOMATICALLY: String = "refresh_timeline_automatically"
    val KEY_SHOW_THREADS_OF_CONVERSATION: String = "show_threads_of_conversation"
    val KEY_MAX_DISTANCE_BETWEEN_DUPLICATES: String = "max_distance_between_duplicates"
    const val MAX_DISTANCE_BETWEEN_DUPLICATES_DEFAULT = 5
    val KEY_SHOW_MY_ACCOUNT_WHICH_DOWNLOADED: String = "show_my_account_which_downloaded"

    // ----------------------------------------------------------
    // Gestures
    val KEY_LONG_PRESS_TO_OPEN_CONTEXT_MENU: String = "long_press_to_open_context_menu"
    val KEY_SEND_IN_EDITOR_BUTTON: String = "send_in_editor_button"

    // ----------------------------------------------------------
    // Attachments
    val KEY_DOWNLOAD_AND_DISPLAY_ATTACHED_IMAGES: String = "show_attached_images"
    val KEY_DOWNLOAD_ATTACHED_VIDEO: String = "download_attached_video"
    val KEY_ATTACH_IMAGES_TO_MY_NOTES: String = "attach_images"
    val KEY_DOWNLOAD_ATTACHMENTS_OVER_WIFI_ONLY: String = "download_attachments_over_wifi_only"
    val KEY_MAXIMUM_SIZE_OF_ATTACHMENT_MB: String = "maximum_size_of_attachment_mb"
    val KEY_MODERN_INTERFACE_TO_SELECT_AN_ATTACHMENT: String = "use_kitkat_media_chooser"

    // ----------------------------------------------------------
    // Syncing
    val KEY_SYNC_FREQUENCY_SECONDS: String = "fetch_frequency"
    private const val SYNC_FREQUENCY_DEFAULT_SECONDS: Long = 180
    val KEY_SYNC_OVER_WIFI_ONLY: String = "sync_over_wifi_only"
    val KEY_SYNC_WHILE_USING_APPLICATION: String = "sync_while_using_application"
    val KEY_SYNC_INDICATOR_ON_TIMELINE: String = "sync_indicator_on_timeline"
    val KEY_SYNC_AFTER_NOTE_WAS_SENT: String = "sync_after_message_was_sent"
    val KEY_DONT_SYNCHRONIZE_OLD_NOTES: String = "dont_synchronize_old_messages"
    val KEY_CONNECTION_TIMEOUT_SECONDS: String = "connection_timeout"
    private const val CONNECTION_TIMEOUT_DEFAULT_SECONDS: Long = 30

    // ----------------------------------------------------------
    // Filters
    val KEY_FILTER_HIDE_NOTES_BASED_ON_KEYWORDS: String = "hide_messages_based_on_keywords"
    val KEY_FILTER_HIDE_REPLIES_NOT_TO_ME_OR_FRIENDS: String = "hide_replies_not_to_me_or_friends"

    // ----------------------------------------------------------
    // Notifications, How to notify
    val KEY_NOTIFICATION_ICON_ALTERNATIVE: String = "notification_icon_alternative"
    val KEY_NOTIFICATION_METHOD_SOUND: String = "notification_sound"

    // ----------------------------------------------------------
    // Storage
    val KEY_USE_EXTERNAL_STORAGE: String = "use_external_storage"

    /** New value for #KEY_USE_EXTERNAL_STORAGE to be confirmed/processed  */
    val KEY_USE_EXTERNAL_STORAGE_NEW: String = "use_external_storage_new"
    val KEY_HISTORY_SIZE: String = "history_size"
    val KEY_HISTORY_TIME: String = "history_time"
    val KEY_MAXIMUM_SIZE_OF_CACHED_MEDIA_MB: String = "maximum_size_of_cached_media_mb"
    val KEY_ENABLE_ANDROID_BACKUP: String = "enable_android_backup"
    val KEY_BACKUP_DOWNLOADS: String = "backup_downloads"
    val KEY_LAST_BACKUP_URI: String = "last_backup_uri"
    val KEY_APP_INSTANCE_NAME: String = "app_instance_name"

    // ----------------------------------------------------------
    // Information
    // (no keys/settings here yet)
    // ----------------------------------------------------------
    // Troubleshooting (logging and debugging)
    val KEY_COMMANDS_QUEUE: String = "commands_queue"

    /** Minimum logging level for the whole application (i.e. for any tag)  */
    val KEY_MIN_LOG_LEVEL: String = "min_log_level"
    val KEY_DEBUGGING_INFO_IN_UI: String = "debugging_info_in_ui"
    val KEY_LOG_NETWORK_LEVEL_MESSAGES: String = "log_network_level_messages"
    val KEY_LOG_EVERYTHING_TO_FILE: String = "log_everything_to_file"
    val KEY_BACKUP_LOG_FILES: String = "backup_log_files"
    // ----------------------------------------------------------
    // Non-UI persistent items ("preferences")
    /** System time when shared preferences were changed  */
    val KEY_PREFERENCES_CHANGE_TIME: String = "preferences_change_time"
    val KEY_DATA_PRUNED_DATE: String = "data_pruned_date"

    /** Version code of last opened application (int)  */
    val KEY_VERSION_CODE_LAST: String = "version_code_last"
    val KEY_BEING_EDITED_NOTE_ID: String = "draft_message_id"
    private const val COLLAPSE_DUPLICATES_DEFAULT_VALUE = true
    const val BYTES_IN_MB = 1000 * 1000

    fun getDontSynchronizeOldNotes(): Long {
        return SharedPreferencesUtil.getLongStoredAsString(KEY_DONT_SYNCHRONIZE_OLD_NOTES, 0)
    }

    fun getConnectionTimeoutMs(): Int {
        return TimeUnit.SECONDS.toMillis(
                SharedPreferencesUtil.getLongStoredAsString(
                        KEY_CONNECTION_TIMEOUT_SECONDS, CONNECTION_TIMEOUT_DEFAULT_SECONDS)).toInt()
    }

    fun setConnectionTimeoutMs(value: Int) {
        val stringValue = TimeUnit.MILLISECONDS.toSeconds(value.toLong()).coerceAtLeast(1).toString()
        SharedPreferencesUtil.putString(KEY_CONNECTION_TIMEOUT_SECONDS, stringValue)
    }

    /**
     * @return the number of seconds between two sync ("fetch"...) actions.
     */
    fun getSyncFrequencySeconds(): Long {
        return SharedPreferencesUtil.getLongStoredAsString(KEY_SYNC_FREQUENCY_SECONDS,
                SYNC_FREQUENCY_DEFAULT_SECONDS)
    }

    fun isSyncOverWiFiOnly(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_SYNC_OVER_WIFI_ONLY, false)
    }

    fun setIsSyncOverWiFiOnly(overWiFi: Boolean) {
        SharedPreferencesUtil.putBoolean(KEY_SYNC_OVER_WIFI_ONLY, overWiFi)
    }

    fun isSyncWhileUsingApplicationEnabled(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_SYNC_WHILE_USING_APPLICATION, true)
    }

    fun isDownloadAttachmentsOverWiFiOnly(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_DOWNLOAD_ATTACHMENTS_OVER_WIFI_ONLY, true)
    }

    fun setDownloadAttachmentsOverWiFiOnly(overWiFi: Boolean) {
        SharedPreferencesUtil.putBoolean(KEY_DOWNLOAD_ATTACHMENTS_OVER_WIFI_ONLY, overWiFi)
    }

    fun isLongPressToOpenContextMenu(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_LONG_PRESS_TO_OPEN_CONTEXT_MENU, false)
    }

    fun getShowAvatars(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_SHOW_AVATARS, true)
    }

    fun getDownloadAndDisplayAttachedImages(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_DOWNLOAD_AND_DISPLAY_ATTACHED_IMAGES, true)
    }

    fun getDownloadAttachedVideo(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_DOWNLOAD_ATTACHED_VIDEO, false)
    }

    fun getShowOrigin(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_SHOW_ORIGIN, false)
    }

    fun isShowSensitiveContent(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_SHOW_SENSITIVE_CONTENT, false)
    }

    fun setShowSensitiveContent(v: Boolean) {
        SharedPreferencesUtil.putBoolean(KEY_SHOW_SENSITIVE_CONTENT, v)
    }

    fun getTapOnATimelineTitleBehaviour(): TapOnATimelineTitleBehaviour? {
        return TapOnATimelineTitleBehaviour.load(
                SharedPreferencesUtil.getString(KEY_TAP_ON_A_TIMELINE_TITLE_BEHAVIOUR, ""))
    }

    fun getActorInTimeline(): ActorInTimeline {
        return ActorInTimeline.load(SharedPreferencesUtil.getString(KEY_ACTOR_IN_TIMELINE, ""))
    }

    fun isShowDebuggingInfoInUi(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_DEBUGGING_INFO_IN_UI, false)
    }

    fun getActionBarTextHomeIconResourceId(): Int {
        return if (SharedPreferencesUtil.getString(KEY_ACTION_BAR_TEXT_COLOR, "")
                == "ActionBarTextBlack") R.drawable.icon_black_24dp else R.drawable.icon_white_24dp
    }

    fun isShowImageAnimations(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_SHOW_IMAGE_ANIMATIONS, true)
    }

    /**
     * @return System time when AndStatus preferences were last time changed.
     * We take into account here time when accounts were added/removed...
     */
    fun getPreferencesChangeTime(): Long {
        return SharedPreferencesUtil.getLong(KEY_PREFERENCES_CHANGE_TIME)
    }

    /**
     * Event: Preferences have changed right now
     * Remember when last changes to the preferences were made
     */
    fun onPreferencesChanged() {
        SharedPreferencesUtil.forget()
        SharedPreferencesUtil.putLong(KEY_PREFERENCES_CHANGE_TIME, System.currentTimeMillis())
        val context: Context? =  MyContextHolder.myContextHolder.getNow().context
        if (context != null && SharedPreferencesUtil.getBoolean(KEY_ENABLE_ANDROID_BACKUP, false)) {
            BackupManager(context).dataChanged()
        }
    }

    fun isCollapseDuplicates(): Boolean {
        return COLLAPSE_DUPLICATES_DEFAULT_VALUE
    }

    fun isShowThreadsOfConversation(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_SHOW_THREADS_OF_CONVERSATION, true)
    }

    fun setShowThreadsOfConversation(v: Boolean) {
        SharedPreferencesUtil.putBoolean(KEY_SHOW_THREADS_OF_CONVERSATION, v)
    }

    fun setOldNotesFirstInConversation(v: Boolean) {
        SharedPreferencesUtil.putBoolean(KEY_OLD_NOTES_FIRST_IN_CONVERSATION, v)
    }

    fun areOldNotesFirstInConversation(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_OLD_NOTES_FIRST_IN_CONVERSATION, false)
    }

    fun isLogEverythingToFile(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_LOG_EVERYTHING_TO_FILE,  MyContextHolder.myContextHolder.getNow().isTestRun)
    }

    fun isRefreshTimelineAutomatically(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_REFRESH_TIMELINE_AUTOMATICALLY, true)
    }

    fun getMaxDistanceBetweenDuplicates(): Int {
        return SharedPreferencesUtil.getIntStoredAsString(KEY_MAX_DISTANCE_BETWEEN_DUPLICATES,
                MAX_DISTANCE_BETWEEN_DUPLICATES_DEFAULT)
    }

    fun getMaximumSizeOfAttachmentBytes(): Long {
        return (Math.max(SharedPreferencesUtil.getLongStoredAsString(KEY_MAXIMUM_SIZE_OF_ATTACHMENT_MB, 5), 1)
                * BYTES_IN_MB)
    }

    val maximumSizeOfCachedMediaBytes: Long get() = SharedPreferencesUtil.getLongStoredAsString(
        KEY_MAXIMUM_SIZE_OF_CACHED_MEDIA_MB, 1000)
        .coerceAtLeast(1) * BYTES_IN_MB

    fun isBackupDownloads(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_BACKUP_DOWNLOADS, false)
    }

    fun isShowMyAccountWhichDownloadedActivity(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_SHOW_MY_ACCOUNT_WHICH_DOWNLOADED, false)
    }

    fun setBeingEditedNoteId(id: Long) {
        SharedPreferencesUtil.putLong(KEY_BEING_EDITED_NOTE_ID, id)
    }

    fun getBeingEditedNoteId(): Long {
        return SharedPreferencesUtil.getLong(KEY_BEING_EDITED_NOTE_ID, 0)
    }

    fun isBackupLogFiles(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_BACKUP_LOG_FILES, false)
    }

    fun isLogNetworkLevelMessages(): Boolean {
        return SharedPreferencesUtil.getBoolean(KEY_LOG_NETWORK_LEVEL_MESSAGES, false)
    }

    fun setLastBackupUri(backupUri: Uri?) {
        SharedPreferencesUtil.putString(KEY_LAST_BACKUP_URI, if (UriUtils.isEmpty(backupUri)) "" else backupUri.toString())
    }

    fun getLastBackupUri(): Uri {
        return UriUtils.fromString(SharedPreferencesUtil.getString(KEY_LAST_BACKUP_URI,
                "content://com.android.externalstorage.documents/tree/primary%3Abackups%2FAndStatus"))
    }

    fun getDeviceBrandModelString(): String {
        return (Build.BRAND + "-" + Build.MODEL).replace(" ".toRegex(), "-")
    }

    fun getAppInstanceName(): String {
        return SharedPreferencesUtil.getString(KEY_APP_INSTANCE_NAME, "")
    }
}
